worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,131 @@
+
+  /**
+   * HikariCP connection pools don't automatically close when IP addresses for a hostname change. This function returns True, iff at
+   * least one of the following conditions is met:
+   *   - IP addresses have changed for the CNAME record used for DNS lookup
+   *   - causeFailoverEveryConnection is set to "true", which is used for testing failover functionality
+   *
+   * @param pool
+   *   the connection pool currently being used
+   * @param dnsIP
+   *   the IP returned when performing a DNS lookup
+   * @return
+   */
+  private def hasIpAddressChanged(pool: Pool, dnsIP: String): Boolean = {
+    causeFailoverEveryConnection == true || pool.ip != dnsIP
   }
 
-  def getConnection = provider.get.getConnection
+  def retryGettingIp[T](n: Int)(fn: => T): T = {
+    try {
+      fn
+    } catch {
+      // Failed to resolve it from JVM
+      case e if n > 1 => retryGettingIp(n - 1)(fn)
+    }
+  }
+
+  val numRetries = 10
+  def getIP: String = {
+    retryGettingIp(numRetries) {
+      // Get the ip address of the hostname. The result is cached in the JVM
+      java.net.InetAddress.getByName(dataSourceHostname).getHostAddress
+    }
+  }
+
+  def getConnection = {
+    if (detectIpAddressFailover && originalJdbcURL != null) {
+      // If the IP has changed, then we know a failover has occurred, and we need to create a new hikari config
+      val newIP: String = getIP
+      if (hasIpAddressChanged(pool, newIP)) {
+        // A failover has occurred, so we lock the pool and swap it out with a new hikari config
+        val (poolToUse, optionalOldPool) = poolLock synchronized {
+          val oldPool = pool
+          val newIP: String = getIP
+          if (hasIpAddressChanged(oldPool, newIP)) {
+            val newPool = new Pool(newIP)
+            pool = newPool
+            (newPool, Some(oldPool))
+          } else {
+            // Already up to date by another thread
+            (oldPool, None)
+          }
+        }
+
+        // Close the previous config
+        optionalOldPool.foreach { old =>
+          // TODO: Get "server.databaseName" from somewhere
+          logger.info(s"Closing DB connection pool for ${originalJdbcURL}: IP changed (${old.ip} -> ${poolToUse.ip}).")
+          old.connectionProvider.get.shutdown()
+        }
+
+        poolToUse.connectionProvider.get.getConnection
+      } else {
+        pool.connectionProvider.get.getConnection
+      }
+    } else {
+      pool.connectionProvider.get.getConnection
+    }
+  }
+
+  private def warmUpCP(connectionPool: HikariCpPoolingConnectionProvider): Unit = {
+    var testConn: Connection = null
+    val start = System.currentTimeMillis
+    while (testConn == null && (System.currentTimeMillis - start) < 2500) {
+      try {
+        testConn = connectionPool.getConnection()
+      } catch {
+        case _: SQLTransientConnectionException => { TimeUnit.MILLISECONDS.sleep(100) } // do nothing
+        case e: Exception                       => throw e
```

Comment:
This is unnecessary. If you don't catch an excetpion in a catch block it will automatically get rethrown